packages/webpack/mixins/build/mixin.core.js

```diff
@@ -81,13 +83,41 @@ class WebpackBuildMixin extends Mixin {
       })
     );
   }
-  diagnose({ diagnoseDuplicatePackages, diagnoseInvalidPackages }) {
-    diagnoseDuplicatePackages('webpack');
-    diagnoseInvalidPackages(__dirname, '@babel/polyfill');
-    diagnoseInvalidPackages(
-      dirname(require.resolve('@babel/polyfill/package.json')),
-      'core-js'
-    );
+  checkHoistedVersion(context, packageName) {
+    const { _workspace } = this.config;
+    const pkgFile = `${packageName}/package.json`;
+    const { version: refVersion } = require(resolve(context, pkgFile));
+    const { version: realVersion } = require(resolve(_workspace, pkgFile));
+    if (refVersion !== realVersion) {
+      return `package ${packageName} must be installed in version ${refVersion}`;
+    }
+  }
+  diagnose({ detectDuplicatePackages, collectResults }) {
+    detectDuplicatePackages('webpack');
+    const warnings = [
+      this.checkHoistedVersion(__dirname, '@babel/polyfill'),
+      this.checkHoistedVersion(
+        dirname(require.resolve('@babel/polyfill/package.json')),
+        'core-js'
+      ),
+    ]
+      .filter(Boolean)
+      .map((message) => `${chalk.yellow('-')} ${message}`);
+
+    const logger = this.getLogger();
+    if (warnings.length) {
+      collectResults(
```

Comment:
When designing that API, this is how I intended it to be used..:

```javascript
collectResults(
  (warnings, { warn, hint }) => {
    if (warnings.length) {
      warn('Invalid package version(s) detected:\n' + warnings
        .map((message) => `${chalk.yellow('-')} ${message}`)
        .join('\n'));
      hint(
        'Fix invalid package version(s):\n' +
          'Please use your package manager of choice (npm or yarn) to ' +
          'install the package versions mentioned above directly into your ' +
          'project. Not doing so will almost certainly lead to hard to ' +
          'debug issues with your application.'
      );
    }
  },
  warnings
);
```

This way, warnings and hints of concerning a specific type of problem will be grouped.

Reply:
Oh, and I forgot to mention that that handler function is rather self-contained, making it a candidate for extraction into a reusable, separate helper function.